src/fswatcher.c

```diff
@@ -0,0 +1,988 @@
+    }
+
+    if (final) {
+        /* we're not going to re-enable, so cleanup */
+        destroy_handle(finf);
+        return;
+    }
+
+    /*
+     * (re)register.
+     */
+    register_watch(key, name, sb);
+}
+
+/*
+ * Worker thread waits here for events, which then get dispatched to
+ * check_and_rearm_event().
+ */
+void *
+wait_for_events(void *arg)
+{
+    (void) arg;
+    struct fileinfo *finf;
+    port_event_t pe;
+
+    while (!port_get(port, &pe, NULL)) {
+        /*
+         * Can add cases for other sources if this
+         * port is used to collect events from multiple sources.
+         */
+        switch (pe.portev_source) {
+        case PORT_SOURCE_FILE:
+            /* Call file events event handler */
+            check_and_rearm_event(0, (char *)pe.portev_user,
+                pe.portev_events, 0);
+            break;
+        default:
+            /*
+             * Something's seriously wrong if we get events with a port source
+             * other than FILE, since that's all we're adding. So abort and hope
+             * there's enough state in the core.
+             */
+            print_error(0, ERR_UNEXPECTED_SOURCE,
+                "event from unexpected source: %s", strerror(errno));
+            abort();
+        }
+
+        /*
+         * The actual work of freeing finf objects is done in this thread so
+         * that there aren't any race conditions while accessing
+         * free_list->fobj.fo_name in the check_and_rearm_event function
+         */
+        mutex_lock(&free_mutex);
+        while (free_list != NULL) {
+            finf = free_list->next;
+            free_handle((struct fileinfo *)free_list);
+            free_list = finf;
+        }
+        mutex_unlock(&free_mutex);
+    }
+    fprintf(stderr, "fswatcher: worker thread exiting\n");
+    fflush(stderr);
+    return (NULL);
+}
+
+/*
+ * Enqueue a finf object to be freed. All this really does is add the structure
+ * to a list that should be freed in the secondary worker thread.
+ *
+ * This is done to prevent race conditions where the finf object would be freed
+ * in the main thread right before an event in the secondary thread would try to
+ * access the name of the object that was going to be stat'd.
+ */
+void
+enqueue_free_finf(struct fileinfo *finf)
+{
+
+    if (finf != NULL) {
+        mutex_lock(&free_mutex);
+        finf->next = (struct fileinfo *)free_list;
+        free_list = finf;
+        finf->prev = NULL;
+        mutex_unlock(&free_mutex);
+    }
+}
+
+/*
+ * Only called from main thread. Attempts to watch pathname.
+ */
+int
+watch_path(char *pathname, uint32_t key, uint64_t start_timestamp)
+{
+    struct fileinfo *finf;
+    /* port is global */
+
+    finf = malloc(sizeof (struct fileinfo));
+    if (finf == NULL) {
+        print_error(key, ERR_CANNOT_ALLOCATE, "failed to allocate memory for "
+            "new watcher errno %d: %s", errno, strerror(errno));
+        /* XXX abort(); ? */
```

Comment:
Probably should abort here so we can figure out where our memory went. It's unlikely we'll be able to make further progress anyway since we're probably leaking.